### PR TITLE
Support setting empty --resolv-conf

### LINKED ISSUE #2638: Implement kubelet flag to set resolv.conf
We have been  experiencing DNS timeouts and `kube-dns` pod restarts. 200+ over four or five days. And we have noticed sporadic "unable to resolve" errors in our apps. One thing we noticed is that we are inheriting the AWS instance's `resolv.conf` in each pod.  It is configured with 3 additional VPC and VPC peering search domains, and we suspect all the additional lookup combinations overload `kube-dns`. We will never need to search these VPC domains for simple host names from inside the cluster (we use FQDNs that will be proxied to the upstream name servers configured in `kube-dns`), so... we'd like to NOT inherit the host's `resolv.conf` at all. We manually updated the `kubelet` config to NOT use  the hosts `resolv.conf`. Simple enough, just added `--resolve-conf=""`.  Things got a lot better. But `kops` has this option commented out in the code, so we can't not do...
```
spec:
  kubelet:
    resolvConf: ""
```
Can we get this uncommented and/or implemented - it would allow us to use `kops` without then having to manually update the `kubelet`. Thanks. @justinsb

### PR DESCRIPTION
**Pull request by austinmoore-**
Adds support for setting the resolv-conf flag on the kubelet through the cluster spec.

If the `spec.kubelet.resolvConf` field is set to an empty string in the cluster spec, the empty string will carry over to the argument like so:
```
--resolv-conf=
```

If the field is not found in the cluster spec, the argument will simply be omitted.

Resolves #2638

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2662)

### CODE REVIEW

File: pkg/flagbuilder/build_flags.go [hunk 1]
```diff
@@ -131,6 +131,12 @@ func BuildFlags(options interface{}) (string, error) {
 				flag = fmt.Sprintf("--%s=%s", flagName, vString)
 			}
 
+		case *string:
```
Comment (chrislovecnm) [inline pkg/flagbuilder/build_flags.go]: Is their a unit test for this, this impacts a lot of code

File: pkg/apis/kops/componentconfig.go [hunk 2]
```diff
@@ -235,9 +235,9 @@ type KubeletConfigSpec struct {
 	// The CIDR to use for pod IP addresses, only used in standalone mode.
 	// In cluster mode, this is obtained from the master.
 	PodCIDR string `json:"podCIDR,omitempty" flag:"pod-cidr"`
-	//// ResolverConfig is the resolver configuration file used as the basis
-	//// for the container DNS resolution configuration."), []
-	//ResolverConfig string `json:"resolvConf"`
+	// ResolverConfig is the resolver configuration file used as the basis
+	// for the container DNS resolution configuration."), []
+	ResolverConfig *string `json:"resolvConf" flag:"resolv-conf"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/componentconfig.go]: What k8s versions does this work on? I am wondering if we need to check versions on this.

Comment (austinmoore-) [inline pkg/apis/kops/componentconfig.go]: I've tested this on 1.6.1. I don't think we need to add a check for this. It seems that the kubelet has supported this flag for a while. The earliest mention of resolv-conf that I could find was [here](https://github.com/kubernetes/kubernetes/commit/1162cfd3b10b3b18aaa8b3b476beea9b1738bdb4) which was from version `v1.2.0-alpha.7`.

File: pkg/flagbuilder/build_flags.go (hunk 1 above)
Comment (austinmoore-) [inline pkg/flagbuilder/build_flags.go]: I'm currently running this code on 1.6.1, and I've tested three cases:
- The field in the cluster spec is exists, but is empty, resulting in `--resolv-conf=`
- The field in the cluster spec exists, and is not empty, resulting in `--resolv-conf=mystring`
- The field in the cluster spec does not exist, resulting in the flag not being built.

I've also added unit tests for these three conditions in `nodeup/pkg/model/kubelet_test.go`, which indirectly test this code. If you think we need more, I can certainly add tests more specific to the `BuildFlags` function. LMK.

Also, while we're talking about this file, I want to get a second pair of eyes on this change here, since I'm still not very familiar with the ins-and-outs of Go, and reflection in Go: https://github.com/kubernetes/kops/pull/2662/files#diff-8f0a190c9dca8ca0e2ed458eb347580fR73
From my tests, this is working fine, and as expected, but I want to make sure that there's not some weird edge case I'm missing.

Comment (chrislovecnm) [inline pkg/flagbuilder/build_flags.go]: @robertojrojas / @justinsb / @kris-nova thoughts?

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

**Comment by chrislovecnm** [comment]
Asked a couple of peeps to take a look at the code block - also since you caught my comment about documentation?

File: pkg/flagbuilder/build_flags.go (hunk 1 above)
Comment (robertojrojas) [inline pkg/flagbuilder/build_flags.go]: I think adding a unit test in the BuildFlags for this scenario is a good idea. Specially, if the 'if' condition was modified to fix a specific problem.

**Comment by austinmoore-** [comment]
Thanks. I'll add those tests, and the documentation.

File: pkg/flagbuilder/build_flags.go (hunk 1 above)
Comment (chrislovecnm) [inline pkg/flagbuilder/build_flags.go]: Does the reflection look ok?

Comment (chrislovecnm) [inline pkg/flagbuilder/build_flags.go]: Austin i would ask that you add some simple tests just for this code. This is a seriously important piece of logic

**Comment by austinmoore-** [comment]
I moved the tests I originally wrote into `buildflags_test.go` because they're more relevant to testing the flagbuilder than the kubelet component. I also added two more test conditions to cover the cases where pointers are not string pointers. Also in the docs, I added a small note about setting empty arguments. Take a look, and let me know what you think @chrislovecnm

File: pkg/flagbuilder/build_flags.go [hunk 3]
```diff
@@ -131,6 +131,12 @@ func BuildFlags(options interface{}) (string, error) {
...
+			if v != nil {
+				vString := fmt.Sprintf("%v", *v)
```
Comment (justinsb) [inline pkg/flagbuilder/build_flags.go]: So actually I think _this_ is fine; we are adding a case here, and we return error on unknown cases, so we know this is additive.   I would prefer us to honor flagEmpty, but we can add that when we first need it.

Comment (justinsb) [inline pkg/flagbuilder/build_flags.go]: We are adding a case here, and we return error on unknown cases, so we know this is additive.

However, because we are no longer flattening `*string` to `string`, the `*string` block needs to be identical to the `string` block.

I think we need to do this through another tag (like `flag-empty`).  Perhaps a tag `flag-include-empty`?  The idea is that even if the flag is empty, we still want to send it.

If we can do that, and the `*string` and `string` blocks are then identical, this then LGTM.

**Comment by chrislovecnm** [comment]
@justinsb we good on this PR now? 

@austinmoore- when you have a chance if you can squash the commits, that would be awesome.  If you want to wait till @justinsb gives you a 👍 no problem.

**Comment by chrislovecnm** [comment]
@justinsb assigning to you since you are the last with some comments.

**Comment by austinmoore-** [comment]
@chrislovecnm Sure thing

**Comment by austinmoore-** [comment]
The Linux Foundation check seems stuck. Is there a way to have it resend the report?

**Comment by stevemcquaid** [comment]
If I can do anything to help move this along, LMK :D

**Comment by chrislovecnm** [comment]
@stevemcquaid waiting on @justinsb for a final review.

File: docs/cluster_spec.md [hunk 4]
```diff
@@ -135,6 +135,15 @@ Will result in the flag `--runtime-config=batch/v2alpha1=true,apps/v1alpha1=true
 
 This block contains configurations for `kubelet`.  See https://kubernetes.io/docs/admin/kubelet/
 
+NOTE: Arguments can be set to empty if the field is included in the spec, and an empty string is passed to it.
```
Comment (justinsb) [inline docs/cluster_spec.md]: This probably should say "Some arguments"

**Comment by justinsb** [comment]
Looks good.  Going to send a followup for some nits so we can get this in!
